contracts/access/TimelockController.sol

```diff
@@ -0,0 +1,294 @@
+
+    /**
+     * @dev Initializes the contract with a given `minDelay`.
+     */
+    constructor(uint256 minDelay) public {
+        _setRoleAdmin(ADMIN_ROLE, ADMIN_ROLE);
+        _setRoleAdmin(PROPOSER_ROLE, ADMIN_ROLE);
+        _setRoleAdmin(EXECUTOR_ROLE, ADMIN_ROLE);
+        _setupRole(ADMIN_ROLE, _msgSender());
+
+        _minDelay = minDelay;
+        emit MinDelayChange(minDelay, 0);
+    }
+
+    /**
+     * @dev Modifier to make a function callable only by a certain role. In
+     * addition to checking the sender's role, `address(0)` 's role is also
+     * considered. Granting a role to `address(0)` is equivalent to enabling
+     * this role for everyone.
+     */
+    modifier onlyRole(bytes32 role) {
+        require(hasRole(role, _msgSender()) || hasRole(role, address(0)), "TimelockController: sender requiers permission");
+        _;
+    }
+
+    /*
+     * @dev Contract might receive/hold ETH as part of the maintenance process.
+     */
+    receive() external payable {}
+
+    /**
+     * @dev Returns whether an operation is pending or not.
+     */
+    function isOperationPending(bytes32 id) public view returns (bool pending) {
+        return _timestamps[id] > _DONE_TIMESTAMP;
+    }
+
+    /**
+     * @dev Returns whether an operation is ready or not.
+     */
+    function isOperationReady(bytes32 id) public view returns (bool ready) {
+        // solhint-disable-next-line not-rely-on-time
+        return _timestamps[id] > _DONE_TIMESTAMP && _timestamps[id] <= block.timestamp;
+    }
+
+    /**
+     * @dev Returns whether an operation is done or not.
+     */
+    function isOperationDone(bytes32 id) public view returns (bool done) {
+        return _timestamps[id] == _DONE_TIMESTAMP;
+    }
+
+    /**
+     * @dev Returns the timestamp at with an operation becomes ready (0 for
+     * unset operations, 1 for done operations).
+     */
+    function viewTimestamp(bytes32 id) public view returns (uint256 timestamp) {
+        return _timestamps[id];
+    }
+
+    /**
+     * @dev Returns the minimum delay for an operation to become valid.
+     */
+    function viewMinDelay() public view returns (uint256 duration) {
+        return _minDelay;
+    }
+
+    /**
+     * @dev Returns the identifier of an operation containing a single
+     * transaction.
+     */
+    function hashOperation(address target, uint256 value, bytes calldata data, bytes32 predecessor, bytes32 salt) public pure returns (bytes32 hash) {
+        return keccak256(abi.encode(target, value, data, predecessor, salt));
+    }
+
+    /**
+     * @dev Returns the identifier of an operation containing a batch of
+     * transactions.
+     */
+    function hashOperationBatch(address[] calldata targets, uint256[] calldata values, bytes[] calldata datas, bytes32 predecessor, bytes32 salt) public pure returns (bytes32 hash) {
+        return keccak256(abi.encode(targets, values, datas, predecessor, salt));
```

Comment:
Is there any way in which the output of `hashOperationBatch` could have a collision with some output of `hashOperation`? That is, are the images of `keccak256(abi.encode(target, value, data, predecessor, salt)` and the batch `abi.encode(targets, values, datas, predecessor, salt)` disjoint?

I don't remember exactly how arrays are encoded, but if I'm not mistaken the only inputs that produce the same encoding in this case are with `target = 0, value = 0, data = ""` and `targets = [], values = [], data = []`, which doesn't look like it could be a problem.

Reply:
As discussed offline to avoid this collision we want to require that the arrays have at least one element.

Reply:
After tests, I noticed that the example above does not cause collision:

```
web3.eth.abi.encodeParameters([
	'address',
	'uint256',
	'bytes',
	'uint256',
	'bytes32',
], [
	constants.ZERO_ADDRESS,
	0,
	'0x',
	constants.ZERO_BYTES32,
	constants.ZERO_BYTES32,
])
→ 0x0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
web3.eth.abi.encodeParameters([
	'address[]',
	'uint256[]',
	'bytes[]',
	'uint256',
	'bytes32',
], [
	[],
	[],
	[],
	constants.ZERO_BYTES32,
	constants.ZERO_BYTES32,
])
→ 0x00000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000c000000000000000000000000000000000000000000000000000000000000000e000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
```

Reply:
Oh I see, that makes sense. We were thinking of the way arrays are encoded in Solidity storage or memory, which is different from the ABI encoding.

Reply:
After some tinkering I came close to a collision but found that for the `bytes` value in the non-batch encoding there will always be a `0x0...0a0` in the third word, whereas in the batch encoding that word will be `0x0...0e0` or greater. Since those two words will always be different, there can't be a collision. In this case we can keep the code as is, I also don't see the need to add the length restriction.